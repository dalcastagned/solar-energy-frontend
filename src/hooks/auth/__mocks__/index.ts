export const userMock = {
  user: 'teste@teste.com',
  roles: [],
  token:
    'eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJBVVJPUkEiLCJhdWQiOiJBVVJPUkEiLCJleHAiOjMwMjc2NTMzNTAsInJvbGVzIjpbIlBBQ1NOR19ESUNPTV9CVUxLREFUQSIsIlBBQ1NOR19ESUNPTV9XQURPIiwiUEFDU05HX0RJQ09NX1FJRE8iLCJQQUNTTkdfQ09ORklHVVJBVElPTiIsIlBBQ1NOR19OT0RFX0NPTlRST0wiLCJQQUNTTkdfU0VSSUVTX01JR1JBVElPTiIsIlBBQ1NOR19TRVJJRVNfUkVNT1ZBTCJdLCJwcmVmZXRjaHJvbGVzIjpbXSwidXNlciI6InVzZXJ0ZXN0ZSJ9.8a_ZdpxfM1_rU-ekCaZnefGFsJVq2_LetC5yBD_80SE',
};

export const loginRequestMock = {
  email: 'usuario@usuario.com',
  password: 'password',
};

