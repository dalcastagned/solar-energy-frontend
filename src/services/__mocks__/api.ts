import axios from 'axios';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

const api = jest.genMockFromModule('axios') as jest.MockedFunction<
  typeof axios
>;

api.create = jest.fn(() => api);

export default api;
