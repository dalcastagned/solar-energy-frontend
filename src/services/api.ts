import axios, { AxiosError, AxiosResponse } from 'axios';

// import { cleanStorage } from '../hooks/auth';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

export type ErrorResponse = {
  response: Error;
  code: string;
};

const responseSuccessHandler = (response: AxiosResponse): AxiosResponse =>
  response;

const forbidden = [401, 403];

const responseErrorHandler = (error: AxiosError): Promise<AxiosError> => {
  const isUnauthorized =
    error.response && forbidden.includes(error.response.status);
  const routers = ['/', '/reset-password'].includes(window.location.pathname);
  if (!routers && isUnauthorized) {
    // cleanStorage();
    window.location.href = '/';
  }

  return Promise.reject(error);
};

const api = axios.create({ baseURL: 'http://localhost:7140/api' });

api.interceptors.response.use(
  response => responseSuccessHandler(response),
  error => responseErrorHandler(error),
);

export default api;
