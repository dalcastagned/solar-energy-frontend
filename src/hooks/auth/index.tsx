import {
  createContext,
  ReactNode,
  useCallback,
  useContext,
  useState,
} from 'react';

import jwt_decode, { JwtPayload } from 'jwt-decode';

import {
  LoginCredentials,
  UpdatePasswordCredentials,
  User,
} from '../../models/user';
import api from '../../services/api';

export type AuthState = {
  user: User;
};

export type AuthContextData = {
  user: User;
  signIn(credentials: LoginCredentials): Promise<void>;
  signOut(): void;
  updatePassword(credentials: UpdatePasswordCredentials): Promise<void>;
  signUp(credentials: LoginCredentials): Promise<void>;
};

export type AuthProviderParams = {
  children: ReactNode;
};

export const cleanStorage = (): void => {
  localStorage.removeItem('user');
};

export const handleToken = (token: string): void => {
  api.defaults.headers.common['X-JWT'] = token;
  api.defaults.headers.common.Authorization = `Bearer ${token}`;
};

const AuthContext = createContext<AuthContextData>({} as AuthContextData);

const AuthProvider = ({ children }: AuthProviderParams): JSX.Element => {
  const [data, setData] = useState<AuthState>(() => {
    const user = localStorage.getItem('user');
    const parsedUser = user ? JSON.parse(user) : null;
    const token = parsedUser?.token;

    if (token && user) {
      const tokenExpirationTime = jwt_decode<JwtPayload>(token).exp;

      const date = new Date().getTime() / 1000;

      if (tokenExpirationTime && tokenExpirationTime < date) {
        cleanStorage();

        return {} as AuthState;
      }

      handleToken(token);

      return { user: parsedUser };
    }
    return {} as AuthState;
  });

  const signIn = useCallback(async ({ email, password }: LoginCredentials) => {
    const response = await api.post<User>('/user/login', {
      email,
      password,
    });

    const { token } = response.data;
    const user = response.data;

    localStorage.setItem('user', JSON.stringify(user));

    handleToken(token);

    setData({ user });
  }, []);

  const updatePassword = useCallback(
    async ({
      user,
      currentPassword,
      newPassword,
    }: UpdatePasswordCredentials) => {
      await api.post<User>(`/user/reset-password/${user}`, {
        currentPassword,
        newPassword,
      });
    },
    [],
  );

  const signUp = useCallback(async ({ email, password }: LoginCredentials) => {
    await api.post<User>('/user/signup', {
      email,
      password,
    });
  }, []);

  const signOut = useCallback(() => {
    cleanStorage();

    setData({} as AuthState);
  }, []);

  return (
    <AuthContext.Provider
      value={{
        user: data.user,
        signIn,
        signOut,
        updatePassword,
        signUp,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

const useAuth = (): AuthContextData => useContext(AuthContext);

export { AuthProvider, useAuth };
