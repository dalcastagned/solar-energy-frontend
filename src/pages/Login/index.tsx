import { useState } from 'react';

import EmailIcon from '@mui/icons-material/Email';
import LockIcon from '@mui/icons-material/Lock';
import { Fade, InputAdornment } from '@mui/material';
import { useFormik } from 'formik';
import { useHistory } from 'react-router-dom';
import * as Yup from 'yup';

import { useAuth } from '../../hooks/auth';
import { LoginCredentials } from '../../models/user';
import { ErrorResponse, handleErrorResponse } from '../../utils/error-response';

import * as S from './styles';

const validationSchema = Yup.object({
  email: Yup.string()
    .email('Entre com um email válido')
    .required('Email é obrigatório'),
  password: Yup.string()
    .min(6, 'Senha precisa ter no mínimo 6 caracteres')
    .max(60, 'Máximo de 60 caracteres')
    .required('Senha é obrigatória'),
});

const Login = (): JSX.Element => {
  const { signIn } = useAuth();
  const [errorMessage, setErrorMessage] = useState('');
  const [loading, setLoading] = useState(false);
  const history = useHistory();

  const handleLogin = async (credentials: LoginCredentials): Promise<void> => {
    try {
      setLoading(true);
      await signIn(credentials);
    } catch (err) {
      const errorMessageToValidate = (
        err as ErrorResponse
      ).response?.data?.toString();

      if (errorMessageToValidate?.includes('Password expired')) {
        history.push(`/reset-password`, { email: credentials.email });
      }
      setErrorMessage(handleErrorResponse(err));
      setLoading(false);
    }
  };

  const onCloseAlert = (): void => {
    setErrorMessage('');
  };

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema,
    onSubmit: values => {
      handleLogin(values);
    },
  });

  return (
    <S.Wrapper>
      <S.ContainerImage />
      <Fade in timeout={1000}>
        <S.ContainerLogin>
          <S.ContainerInfo>
            <img src="/img/logo.png" alt="Logo" />
            <h1>Seja bem vindo</h1>
          </S.ContainerInfo>
          <S.Form onSubmit={formik.handleSubmit}>
            {errorMessage && (
              <S.AlertStyled severity="error" onClose={onCloseAlert}>
                {errorMessage}
              </S.AlertStyled>
            )}
            <S.TextFieldStyled
              id="email"
              name="email"
              label="Email"
              InputProps={{
                placeholder: 'exemplo@exemplo.com',
                startAdornment: (
                  <InputAdornment position="start">
                    <EmailIcon />
                  </InputAdornment>
                ),
              }}
              value={formik.values.email}
              onChange={formik.handleChange}
              error={formik.touched.email && Boolean(formik.errors.email)}
              helperText={formik.touched.email && formik.errors.email}
            />
            <S.TextFieldStyled
              id="password"
              name="password"
              label="Senha"
              type="password"
              InputProps={{
                placeholder: 'Digite sua senha',
                startAdornment: (
                  <InputAdornment position="start">
                    <LockIcon />
                  </InputAdornment>
                ),
              }}
              value={formik.values.password}
              onChange={formik.handleChange}
              error={formik.touched.password && Boolean(formik.errors.password)}
              helperText={formik.touched.password && formik.errors.password}
            />
            <S.ButtonStyled
              color="primary"
              variant="contained"
              fullWidth
              type="submit"
              disabled={loading}
            >
              Entrar
            </S.ButtonStyled>
          </S.Form>
          <S.LinkStyled href="/signup">
            Ainda não possui uma conta? Cadastre-se!
          </S.LinkStyled>
        </S.ContainerLogin>
      </Fade>
    </S.Wrapper>
  );
};

export default Login;
