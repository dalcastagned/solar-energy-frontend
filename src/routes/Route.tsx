import { ComponentType } from 'react';

import {
  Redirect,
  Route as ReactDOMRoute,
  RouteComponentProps,
  RouteProps as ReactDOMRouteProps,
} from 'react-router-dom';
import { whereRedirect } from 'utils/tests/where-redirect';

import { useAuth } from '../hooks/auth';

export type RouteProps = {
  isPrivate?: boolean;
  Component: ComponentType<RouteComponentProps>;
  roles?: string[];
} & ReactDOMRouteProps;

const Route = ({
  isPrivate = false,
  Component,
  ...props
}: RouteProps): JSX.Element => {
  const { user } = useAuth();

  const render = (params: RouteComponentProps): React.ReactNode => {
    const isAuthenticated = isPrivate === !!user;

    if (!isAuthenticated) {
      return (
        <Redirect
          to={{
            pathname: whereRedirect(isPrivate),
            state: { from: params.location },
          }}
        />
      );
    }

    return <Component {...params} />;
  };

  return <ReactDOMRoute {...props} render={render} />;
};

export default Route;
