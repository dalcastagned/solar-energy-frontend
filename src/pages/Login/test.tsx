import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Login from '.';
import { AuthProvider } from '../../hooks/auth';
import Route from '../../routes/Route';
import api from '../../services/api';
import { renderWithRoute } from '../../utils/tests/helpers';

describe('<Login />', () => {
  it('should render the main', () => {
    const { container } = renderWithRoute(<Login />);
    expect(screen.getByText('Entrar')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the validation messages when login without username and password', async () => {
    renderWithRoute(<Login />);

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      expect(screen.getByText('Email é obrigatório')).toBeInTheDocument();
      expect(screen.getByText('Senha é obrigatória')).toBeInTheDocument();
    });
  });

  it('must render validation messages when password contain more than 60 characters', async () => {
    renderWithRoute(<Login />);

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, Array(61).fill('a').join(''));

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      expect(screen.getByText('Máximo de 60 caracteres')).toBeInTheDocument();
    });
  });

  it('should render alert when user is not authorized', async () => {
    jest.spyOn(api, 'post').mockImplementation(async () => {
      const error = { response: { status: 401 } };
      throw error;
    });

    renderWithRoute(
      <AuthProvider>
        <Login />
      </AuthProvider>,
    );

    const userInput = screen.getByLabelText('Email');
    userEvent.type(userInput, 'usuario@usuario.com');

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, '123456');

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      expect(
        screen.getByText('Usuário ou senha inválidos.'),
      ).toBeInTheDocument();
    });
  });

  it('should render alert when Password Expired', async () => {
    jest.spyOn(api, 'post').mockRejectedValue({
      response: { status: 401, data: 'Password expired' },
    });

    const resetPage = (): JSX.Element => <h1>Alterar senha</h1>;

    renderWithRoute(
      <AuthProvider>
        <Route Component={Login} />
        <Route Component={resetPage} path="/reset-password" />
      </AuthProvider>,
    );

    const userInput = screen.getByLabelText('Email');
    userEvent.type(userInput, 'usuario@usuario.com');

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, '123456');

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      expect(screen.getByText('Alterar senha')).toBeInTheDocument();
    });
  });

  it('should render alert when server is down', async () => {
    jest.spyOn(api, 'post').mockImplementation(async () => {
      const error = { code: 'ECONNABORTED' };
      throw error;
    });

    renderWithRoute(
      <AuthProvider>
        <Login />
      </AuthProvider>,
    );

    const userInput = screen.getByLabelText('Email');
    userEvent.type(userInput, 'usuario@usuario.com');

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, '123456');

    const loginButton = screen.getByRole('button', { name: 'Entrar' });

    userEvent.click(loginButton);

    await waitFor(async () => {
      expect(
        screen.getByText(
          'O servidor parece estar indisponível. Verifique sua conexão.',
        ),
      ).toBeInTheDocument();
    });
  });

  it('should render alert with server error message', async () => {
    jest.spyOn(api, 'post').mockImplementation(async () => {
      const error = { response: { status: 500 } };
      throw error;
    });

    renderWithRoute(
      <AuthProvider>
        <Login />
      </AuthProvider>,
    );

    const userInput = screen.getByLabelText('Email');
    userEvent.type(userInput, 'usuario@usuario.com');

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, '123456');

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      expect(
        screen.getByText('Ocorreu um erro no servidor.'),
      ).toBeInTheDocument();
    });
  });

  it('should close alert when clicked on closeButton', async () => {
    jest.spyOn(api, 'post').mockImplementation(async () => {
      const error = { response: undefined };
      throw error;
    });

    renderWithRoute(
      <AuthProvider>
        <Login />
      </AuthProvider>,
    );

    const userInput = screen.getByLabelText('Email');
    userEvent.type(userInput, 'usuario@usuario.com');

    const passwordInput = screen.getByLabelText('Senha');
    userEvent.type(passwordInput, '123456');

    const loginButton = screen.getByText('Entrar');

    userEvent.click(loginButton);

    await waitFor(() => {
      const closeAlert = screen.getByLabelText('Close');
      userEvent.click(closeAlert);

      expect(screen.queryByRole('alert')).not.toBeInTheDocument();
    });
  });
});
