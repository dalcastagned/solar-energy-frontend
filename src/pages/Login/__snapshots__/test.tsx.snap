// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render the main 1`] = `
<div
  class="css-f9gpob"
>
  <div
    class="css-jtmvlx"
  />
  <div
    class="css-78rvji"
    style="opacity: 1; webkit-transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="css-1tyosks"
    >
      <img
        alt="Logo"
        src="/img/logo.png"
      />
      <h1>
        Seja bem vindo
      </h1>
    </div>
    <form
      class="css-lybupd"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-5uzda4-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="email"
          id="email-label"
        >
          Email
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EmailIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            placeholder="exemplo@exemplo.com"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1in441m"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-5uzda4-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="password"
          id="password-label"
        >
          Senha
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LockIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1in441m"
            >
              <span>
                Senha
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-1j4k5xr-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Entrar
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1qk869o-MuiTypography-root-MuiLink-root"
      href="/signup"
    >
      Ainda não possui uma conta? Cadastre-se!
    </a>
  </div>
</div>
`;
